src/oemof/solph/flows/_non_convex_invest_flow.py

```diff
@@ -0,0 +1,860 @@
+
+class NonConvexInvestFlow(Flow):
+    r"""
+    Investment flow with a binary variable that states whether
+    it is active or not.
+
+    Parameters
+    ----------
+    startup_costs : numeric (iterable or scalar)
+        Costs associated with a start of the flow (representing a unit).
+    shutdown_costs : numeric (iterable or scalar)
+        Costs associated with the shutdown of the flow (representing a unit).
+    minimum_uptime : numeric (1 or positive integer)
+        Minimum time that a flow must be greater then its minimum flow after
+        startup. Be aware that minimum up and downtimes can contradict each
+        other and may lead to infeasible problems.
+    minimum_downtime : numeric (1 or positive integer)
+        Minimum time a flow is forced to zero after shutting down.
+        Be aware that minimum up and downtimes can contradict each
+        other and may to infeasible problems.
+    maximum_startups : numeric (0 or positive integer)
+        Maximum number of start-ups.
+    maximum_shutdowns : numeric (0 or positive integer)
+        Maximum number of shutdowns.
+    initial_status : numeric (0 or 1)
+        Integer value indicating the status of the flow in the first time step
+        (0 = off, 1 = on). For minimum up and downtimes, the initial status
+        is set for the respective values in the edge regions e.g. if a
+        minimum uptime of four timesteps is defined, the initial status is
+        fixed for the four first and last timesteps of the optimization period.
+        If both, up and downtimes are defined, the initial status is set for
+        the maximum of both e.g. for six timesteps if a minimum downtime of
+        six timesteps is defined in addition to a four timestep minimum uptime.
+    positive_gradient : :obj:`dict`, default: `{'ub': None, 'costs': 0}`
+        A dictionary containing the following two keys:
+         * `'ub'`: numeric (iterable, scalar or None), the normed *upper
+           bound* on the positive difference (`flow[t-1] < flow[t]`) of
+           two consecutive flow values.
+         * `'costs``: numeric (scalar or None), the gradient cost per
+           unit.
+    negative_gradient : :obj:`dict`, default: `{'ub': None, 'costs': 0}`
+        A dictionary containing the following two keys:
+          * `'ub'`: numeric (iterable, scalar or None), the normed *upper
+            bound* on the negative difference (`flow[t-1] > flow[t]`) of
+            two consecutive flow values.
+          * `'costs``: numeric (scalar or None), the gradient cost per
+            unit.
+    """
+
+    def __init__(
+        self,
+        startup_costs=None,
+        shutdown_costs=None,
+        activity_costs=None,
+        inactivity_costs=None,
+        minimum_uptime=None,
+        minimum_downtime=None,
+        maximum_startups=None,
+        maximum_shutdowns=None,
+        initial_status=0,
+        positive_gradient=None,
+        negative_gradient=None,
+        **kwargs,
```

Comment:
As `NonConvexInvestmentFlow` inherits from `Flow` and `Flow` itself uses `**kwargs`, we then might as well define explicitely the argument in `Flow`, no? Couldn't this be tackled in a separate PR?

Reply:
I understand the benefit of explicit by getting rid of `**kwargs`, however wouldn't it make maintenance of inherited classes more cumbersome ? Moreover if typos are caught, then errors in the default values of the arguments in a child class are not caught (a solution for that would be to only pass the argument to the parent class if they don't have the default value of the child class, that way only the default value of the parent class rules.)

Reply:
You are right, this can be done separately, as the Flow is not updated yet.

(Maybe, this PR is not the right place to discuss about the pros and cons of **kwags. We came to the conclusion to user friendlyness is the stronger argument in the current state of the package. So, all expected arguments are now made explicit. If you want additional arguments, you can still use them, but you have to hand them as a dict to the argument "options".)